src/ensembl/src/shared/hooks/useApiService.ts

```diff
@@ -0,0 +1,113 @@
+import { useEffect, useReducer, Reducer } from 'react';
+
+import apiService, { FetchOptions, APIError } from 'src/services/api-service';
+
+import { LoadingState } from 'src/shared/types/loading-state';
+
+type Params = FetchOptions & {
+  endpoint: string;
+  isAbortable?: boolean;
+};
+
+type StateAtLoading = {
+  loadingState: LoadingState.LOADING;
```

Comment:
Just out of curiosity, are there any advantages in using the `LoadingState` enum for the loadingState instead of `boolean`?

Reply:
I don't think `boolean` is suited for this as we also have types for success and error which will be of different `loadingState` type.

Reply:
It's a difficult question, because it can have several answers.

On a purely practical level, the answer would be that the `isLoading` boolean in `useQuery` clearly works. When writing a components, one can first check `isLoading`, then, if it's false, assume that either the `data` or the `error` will have a value, and so check the `error` field and the `data` field. It works.

The answer on a more abstract level, though, is that the `isLoading` boolean does not tell you anything about whether your request has succeeded or failed. It does not allow you to infer the values of your `data` and error fields:

success:
```js
{
  isLoading: false,
  data: { foo: 'bar' },
  error: null
}
```

failure:
```js
{
  isLoading: false,
  data: null,
  error: { message: 'oh no!' }
}
```

🤷‍♂️ 
```js
{
  isLoading: false,
  data: null,
  error: null
}
```

🙅‍♂️ (but no guarantee in code to guard against this)
```js
{
  isLoading: false,
  data: { foo: 'bar' },
  error: { message: 'oh no!' }
}
```

The last combination is clearly invalid; the one before last is confusing (either invalid or initial).

Using an enum field as a flag allows you to type different combinations of fields and to eliminate impossisble combinations; so from the perspective of type guarantees, it's a definite win.

A more robust and formal alternative is using a state machine; but I don't know whether we are ready to go quite as far.

There are a number of blog posts on this topic (e.g. [one by Kent C. Dodds](https://kentcdodds.com/blog/stop-using-isloading-booleans), or [one by ThoughtBot](https://thoughtbot.com/blog/booleans-and-enums))

Reply:
Cool. Thanks for that.

Do you think we need an option to `skip`?

Reply:
> Do you think we need an option to `skip`?

Good question! `skip` will complicate the matters slightly, because it will introduce a new loading state possibility — where no request has been made:

```
{
  loadingState: LoadingState.NOT_REQUESTED;
  data: null;
  error: null;
}
```
I didn't have a need for it, but probably a good thing to add it just for completeness' sake.